common/OpTestSSH.py

```diff
@@ -0,0 +1,180 @@
+               + " -o'RSAAuthentication=no' -o 'PubkeyAuthentication=no'"
+               + " -o 'UserKnownHostsFile=/dev/null' "
+               + " -o 'StrictHostKeyChecking=no'"
+           )
+
+        print cmd
+        consoleChild = OPexpect.spawn(cmd,logfile=self.logfile,
+                failure_callback=set_system_to_UNKNOWN,
+                failure_callback_data=self.system)
+        self.state = ConsoleState.CONNECTED
+        self.console = consoleChild
+        # Users expecting "Host IPMI" will reference console.sol so make it available
+        self.sol = self.console
+        return consoleChild
+
+    def get_console(self):
+        if self.state == ConsoleState.DISCONNECTED:
+            self.connect()
+
+        count = 0
+        while (not self.console.isalive()):
+            print '# Reconnecting'
+            if (count > 0):
+                time.sleep(1)
+            self.connect()
+            count += 1
+            if count > 120:
+                raise "SSH: not able to get console"
+
+        return self.console
+
+    def run_command(self, command, timeout=60):
+        console = self.get_console()
+
+        if self.prompt:
+            prompt = self.prompt
+        else:
+            console.sendline('PS1=[console-pexpect]\#')
+            console.expect("\n") # from us, because echo
+            prompt = "\[console-pexpect\]#$"
+
+        rc = console.expect([prompt], timeout)
+        output = console.before
+
+        console.sendline(command)
+        console.expect("\n") # from us
+        rc = None
+        output = None
+        exitcode = None
+        try:
+            rc = console.expect([prompt], timeout)
+            output = console.before
+            console.sendline("echo $?")
+            console.expect("\n") # from us
+            rc = console.expect([prompt], timeout)
+            exitcode = int(console.before)
+        except pexpect.TIMEOUT as e:
+            print e
+            print "# TIMEOUT waiting for command to finish."
+            print "# Attempting to control-c"
+            try:
+                console.sendcontrol('c')
+                rc = console.expect(["\[console-pexpect\]#$"], 10)
+                if rc == 0:
+                    raise CommandFailed(command, "TIMEOUT", -1)
+            except pexpect.TIMEOUT:
+                print "# Timeout trying to kill timed-out command."
+                print "# Failing current command and attempting to continue"
+                self.terminate()
+                raise CommandFailed("ssh -p " + self.port, "timeout", -1)
+            raise e
+
+        if rc == 0:
+            res = output
+            res = res.splitlines()
+            self.terminate()
+            if exitcode != 0:
+                raise CommandFailed(command, res, exitcode)
+            return res
+        else:
+            res = console.before
+            res = res.split(command)
+            self.terminate()
+            return res[-1].splitlines()
```

Comment:
And also i would like you to add some checks in HostLogin testcase to run different commands in bmc and host console respectively. so that we can test `HostLogin` itself individually instead of testing entire test suite.
```        
        self.system = conf.system()
        self.bmc = conf.bmc()
        self.bmc.ssh.run_command("dmesg")
        self.bmc.ssh.run_command("ls")
        self.bmc.ssh.run_command("cat /etc/os-release")
        self.bmc.ssh.run_command("dmesg")
        self.system.host_console_login()
        self.system.host_console_unique_prompt()
        l_con = self.system.sys_get_ipmi_console()
        l_con.run_command("dmesg")
        l_con.run_command("cat /sys/firmware/opal/msglog")
        l_con.run_command("dmesg")
```
Kind of unit tests for run_command API, Please add different commands in that and test for correctness of the API.

Reply:
Testing the BMC ssh console in `HostLogin` sounds like the wrong location  - should we instead make a `testcases/BMCLogin.py`?

Reply:
@sammj Sure, that makes sense, have two different tests for HostLogin and BMCLogin. And make sure you execute enough number of commands in those two tests, to test `run_command` API.